Modules/_ssl.c

```diff
@@ -2251,6 +2251,17 @@ PySSL_dealloc(PySSLSocket *self)
     PyTypeObject *tp = Py_TYPE(self);
     PyObject_GC_UnTrack(self);
     if (self->ssl) {
+        // If we free the SSL socket object without having called SSL_shutdown,
+        // OpenSSL will invalidate the linked SSL session object. While this
+        // behavior is strictly RFC-compliant, it makes session reuse less
+        // likely and it would also break compatibility with older stdlib
+        // versions (which used an ugly workaround of duplicating the
+        // SSL_SESSION object).
+        // Therefore, we ensure the socket is marked as shutdown in any case.
+        //
+        // See elaborate explanation at
+        // https://github.com/python/cpython/pull/123249#discussion_r1766164530
+        SSL_set_shutdown(self->ssl, SSL_SENT_SHUTDOWN);
```

Comment:
This probably should be either:
```
SSL_set_shutdown(self->ssl, SSL_SENT_SHUTDOWN | SSL_RECEIVED_SHUTDOWN);
```
or:
```
SSL_set_shutdown(self->ssl, SSL_SENT_SHUTDOWN | SSL_get_shutdown(self->ssl));
```

That way if you have already received `close_notify`, you don't clear the bit. I mean, you're immediately destroying the `SSL` anyway, but convincing the library it hasn't received something it has actually received is a little odd. :)

Reply:
(Also because I had put an `assert()` into BoringSSL that callers don't clear bits, because that's weird. 😉 )